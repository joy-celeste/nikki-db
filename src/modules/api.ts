import { HttpRequestException } from './errors';
import database from './database';
import { API_CONSTANTS } from './constants';

export const fetchItemData = async (itemId: number): Promise<any> =>
  database.collection(API_CONSTANTS.CLOTHES).doc(itemId.toString())
    .get()
    .then((response) => {
      if (!response) {
        console.log('No such document!');
        return null;
      }
      console.log('Retrieved data from database.');
      return response.data();
    })
    .catch((error) => {
      console.log(new HttpRequestException(error, 'Error getting response.'));
    });

export {};
