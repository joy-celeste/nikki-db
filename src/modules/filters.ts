import SearchIndex, { SUITS_BOOST_TERM } from './search';

export type Operator = 'and' | 'or';
export const PLEASE_CREATE_A_FILTER = 'Please create a filter.';
export const PLEASE_FILL_ALL_VALUES = 'Please fill all values.';
export const SEARCH_BUTTON_LABEL = 'Search';

export class FilterSet {
  id: string;
  filters: Filter[];
  operator: Operator;

  constructor(input?: FilterSet) {
    this.id = `filterSet-${Math.random().toString(36).substring(7)}`;
    this.filters = input?.filters ?? [];
    this.operator = input?.operator ?? 'and';
    this.addFilter(new Filter());
  }

  allFiltersValid() {
    return this.filters.every((filter) => filter.isValid());
  }

  generateSubmitMessage() {
    if (this.filters.length === 0) {
      return PLEASE_CREATE_A_FILTER;
    } if (!this.allFiltersValid()) {
      return PLEASE_FILL_ALL_VALUES;
    }
    return SEARCH_BUTTON_LABEL;
  }

  setOperator(operator: Operator) {
    this.operator = operator;
  }

  addFilter(filter: Filter) {
    this.filters.push(filter);
  }

  removeFilter(filter: Filter) {
    this.filters = this.filters.filter((f) => filter.id !== f.id);
  }

  toString() {
    return this.filters.map((f) => f.toString()).join(' ');
  }

  intersection(array: any[][]): any[] {
    return array.reduce((a, b) => b.filter(Set.prototype.has.bind(new Set(a))));
  }

  filtersWithAny() {
    return this.filters.filter((filter) => filter.filterType === 'select' && filter.selectType === 'any');
  }

  filtersWithoutAny() {
    return this.filters.filter((filter) => filter.filterType !== 'select' || filter.selectType !== 'any');
  }

  search(index: SearchIndex) {
    const withAny: Filter[] = this.filtersWithAny();

    // Without any multi-select filters with 'any' selectType ==========================================
    if (!withAny.length) {
      if (this.filters.length === 1 || this.operator === 'and') {
        const searchTerm = `${this.toString()} ${SUITS_BOOST_TERM}`;
        return index.searchWithTerm(searchTerm);
      }
      return this.filters.flatMap((filter: any) => index.searchWithTerm(filter.toString()));
    }

    // With any multi-select filters with 'any' selectType =============================================
    if (this.filters.length === 1) {
      return this.filters[0].search(index);
    }
    const withoutAny: Filter[] = this.filtersWithoutAny();
    const withoutAnySearchTerm = `${withoutAny.map((f) => f.toString()).join(' ')} ${SUITS_BOOST_TERM}`;
    const withoutAnyResults = withoutAny.length ? index.searchWithTerm(withoutAnySearchTerm) : [];

    if (this.operator === 'and') {
      const withAnyResults = withAny.map((filter) => filter.search(index));
      const intersection = this.intersection(withAnyResults);
      return withoutAnyResults.length 
        ? intersection.filter((res) => withoutAnyResults.includes(res))
        : intersection;
    } else {
      const withAnyResults = withAny.flatMap((filter) => filter.search(index));
      return withAnyResults.concat(withoutAnyResults);
    }
  }
}

export type FilterType = 'userInput' | 'select' | 'checkbox';
export type SelectType = 'any' | 'only';

export class Filter {
  id: string;
  filterType: FilterType;
  filterValue: string;
  userInputValue: string;
  userInputContains: boolean;
  selectType: SelectType;
  selections: string[] | string;
  checkboxIsChecked: boolean;

  constructor(input?: Filter) {
    this.id = `filter-${Math.random().toString(36).substring(7)}`;
    this.filterType = input?.filterType ?? 'userInput';
    this.filterValue = input?.filterValue ?? 'name';
    this.userInputValue = input?.userInputValue ?? '';
    this.userInputContains = input?.userInputContains ?? true;
    this.selectType = input?.selectType ?? 'only';
    this.selections = input?.selections ?? [];
    this.checkboxIsChecked = input?.checkboxIsChecked ?? true;
  }

  setFilterType(filterType: FilterType) { this.filterType = filterType; }
  setFilterValue(filterValue: string) { this.filterValue = filterValue; }
  setUserInputValue(userInputValue: string) { this.userInputValue = userInputValue; }
  setUserInputContains(userInputContains: boolean) { this.userInputContains = userInputContains; }
  setSelectType(selectType: SelectType) { this.selectType = selectType; }
  setSelection(selection: string[]) { this.selections = selection; }
  setCheckboxIsChecked(checkboxIsChecked: boolean) { this.checkboxIsChecked = checkboxIsChecked; }

  isValid(): boolean {
    switch (this.filterType) {
      case 'userInput':
        return this.userInputValue.length > 0;
      case 'checkbox':
        return true;
      case 'select':
        return Array.isArray(this.selections) ? !!this.selections.length : !!this.selections;
    }
  }

  toString() {
    if (this.isValid()) {
      switch (this.filterType) {
        case 'userInput':
          const containsPrefix = this.userInputContains ? '+' : '-';
          const sanitizedString = this.userInputValue.split(' ').join('_').toLowerCase();
          return `${containsPrefix}${this.filterValue}:*_${sanitizedString}_*`;
        case 'checkbox':
          return `+${this.filterValue}:${this.checkboxIsChecked}`;
        case 'select':
          if (this.selectType === 'only') {
            return `+${this.filterValue}:${this.selections}`;
          }
      }
    } else {
      return ''
    }
  }

  search(index: SearchIndex, maxResultsEach?: number) {
    if (this.filterType === 'select' && this.selectType === 'any' && Array.isArray(this.selections)) {
      return this.selections.flatMap((s) => index.searchWithTerm(`+${this.filterValue}:${s}`, maxResultsEach));
    }
  }
}