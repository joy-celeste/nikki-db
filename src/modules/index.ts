import { combineReducers } from 'redux';
import { userReducer } from './user';
import { productsReducer } from './products';

export const rootReducer = combineReducers({
  user: userReducer,
  products: productsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;