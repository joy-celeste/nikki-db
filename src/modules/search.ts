import lunr, { Index } from 'lunr';
import { AnyAction } from 'redux';
import searchIndexData from '../data/search_index.json';
import refToSearchResult from '../data/ref_to_search_result.json';
import { ACTION_CONSTANTS, ITEM_SUFFIX } from './constants';
import { RootState } from '.';
import { ItemId } from './data';

export const MAX_RESULTS = 150;

export default class SearchIndex {
  index: Index;

  constructor() {
    this.createIndex();
  }

  /**
   * Process the data and create a search index.
   * NOTE: Stemming is disabled :)
   */
  createIndex(): SearchIndex {
    this.index = lunr(function options() {
      this.pipeline.remove(lunr.stemmer);
      this.searchPipeline.remove(lunr.stemmer);

      this.ref('id');
      this.field('name');
      this.field('spec');
      this.field('rare');
      this.field('posed');
      this.field('tag1');
      this.field('tag2');
      this.field('depth');
      this.field('subtype');
      this.field('genre');
      this.field('isSuit');

      searchIndexData.forEach((doc) => {
        this.add(doc);
      }, this);
    });
    return this;
  }

  /**
   * Looks up the item within the search index.
   * @param searchTerm The search input.
   * @param maxResults The max number of results to return.
   */
  search(searchTerm: string, maxResults = MAX_RESULTS): string[] {
    const output: string[] = [];
    this.index.search(searchTerm).some((result) => {
      output.push(result.ref);
      return output.length === maxResults;
    });
    return output;
  }
}

export type SearchResult = {
  name: string;
  iconId?: ItemId;
  contents: ItemId[];
  posed?: boolean;
  variation?: string;
  isSuit?: boolean
};

export type SearchState = {
  index: SearchIndex;
  results: SearchResult[];
};

export type RefToSearchData = Record<string, SearchData>;

const refToData: RefToSearchData = JSON.parse(JSON.stringify(refToSearchResult));

interface SearchData {
  name: string,
  iconId: ItemId,
  contents: ItemId[],
  posed?: boolean,
  variation?: string,
  isSuit?: boolean
}

const initialState: SearchState = {
  index: new SearchIndex(),
  results: null,
};

// ACTIONS
const updateSearchResults = (searchResults: SearchResult[]): AnyAction => ({
  type: ACTION_CONSTANTS.SEARCH_UPDATE_RESULTS,
  payload: searchResults,
});

// USE-CASE
/* istanbul ignore next */ /* branch not passing coverage check for MAX_RESULT? */
export const searchName = (searchTerm: string, maxResults: number = MAX_RESULTS) =>
  async(dispatch: Function, getState: () => RootState): Promise<void> => {
    const searchState = getState().search;
    const initialResults = searchState.index.search(searchTerm, maxResults);
    const parsedResults: SearchResult[] = initialResults.flatMap((key: string) => {
      const suitData = refToData[key];
      let displayName = suitData?.name;

      if (suitData?.isSuit) {
        if (suitData?.posed) {
          displayName += ITEM_SUFFIX.POSED_SUIT;
        } else {
        displayName += ITEM_SUFFIX.SUIT;
        }
      }

      if (suitData?.variation) {
        return {
          name: displayName,
          iconId: suitData?.iconId,
          contents: suitData?.contents,
          posed: suitData?.posed,
          variation: suitData?.variation,
          isSuit: suitData?.isSuit,
        };
      }
      return {
        name: displayName,
        iconId: suitData?.iconId,
        contents: suitData?.contents,
      };
    });
    dispatch(updateSearchResults(parsedResults));
  };

// REDUCER
export function searchReducer(
  state = initialState,
  action: AnyAction,
): SearchState {
  switch (action.type) {
    case ACTION_CONSTANTS.SEARCH_UPDATE_RESULTS:
      return {
        ...state,
        results: action.payload,
      };
    default:
      return state;
  }
}
