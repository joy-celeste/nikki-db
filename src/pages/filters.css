:root {
  --filter-input-height: 38px;
  --filter-toggle-height: 30px;
}

#filterToggle table {
    width: 100%;
    table-layout: fixed;
    border-radius: var(--border-radius);
    border-collapse: separate;
    border: 1px solid var(--pink);
    margin-bottom: var(--padding);
    background-color: var(--purple-light);
    color: var(--pink);
}

#filterToggle td {
    height: var(--filter-toggle-height);
    text-align: center;
    vertical-align: middle;
}

#filterToggle #selected {
    color: white;
    background-color: var(--pink);
    border-radius: var(--border-radius);
}

#filterToggle a {
    display: block;
    width: 100%;
    border-radius: var(--border-radius);
}

/* ==================== */

#empty-results {
    text-align: center;
}

/* ==================== */

.advancedFilter, .advancedFilters, .simpleSearch {
	width: 100%;
	flex-basis: 100%;
}

.advancedFilter {
	display: flex;
}

.advancedSearch, .simpleSearch {
	width: 100%;
	flex-wrap: wrap;
	position: relative;
	padding-bottom: 10px;
}

.advancedFilterToolbar {
	width: 100%;
	display: flex;
	justify-content: space-between;
	padding-top: 5px;
}

.simpleSearch input, .advancedFilter input, .advancedFilterToolbar input {
    padding: 8px;
    border-radius: 5px;
    border: 1px solid #cdcdcd;
}

.advancedFilterFirst, .advancedFilterFourth, .maxResultsLabel {
	padding-top: 10px;
	padding-bottom: 10px;
	max-height: 17px;
	min-width: 110px;
	max-width: 160px;
	text-align: center;
}

.sortFilter {
	min-width: 150px;
}

.advancedFilterInput input, .maxResultsField input {
    height: 38px;
	width: 100%;
}

.simpleSearch {
	max-width: 100%;
}

.searchbar input {
    height: var(--filter-input-height);
	width: 80%;
}

input.searchbar-submit {
	width: 20%;
}

.searchInputText {
	display: flex;
	justify-content: space-between;
}

.searchInputText b{
	position: relative;
	top: 10px;
}


