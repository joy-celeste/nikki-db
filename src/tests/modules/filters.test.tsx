import { combineReducers, Store } from 'redux';
import { createSeedFunction, createStoreWithMiddleware } from '../helpers';
import { Filter, FilterSet, } from '../../modules/filters';
import { searchReducer, updateFilterSet, updateSearchString, searchInventory, updateMaxResults, setAdvancedSearch } from '../../modules/search';
import { RootState } from '../../modules';

const mockMath = Object.create(global.Math);
mockMath.random = createSeedFunction(42);
global.Math = mockMath;

describe('FilterSet', () => {
  let filterSet: FilterSet;

  beforeEach(() => {
    filterSet = new FilterSet();
  });

  test('Initializes to with correct default data', () => {
    expect(filterSet.operator).toBe('and');
    expect(filterSet.filters.length).toBe(1);
    expect(filterSet.id.substring(0,10)).toBe('filterSet-')
    expect(filterSet).toMatchSnapshot();
  });

  test('Correctly changes operator', () => {
    expect(filterSet.operator).toBe('and');
    filterSet.setOperator('or');
    expect(filterSet.operator).toBe('or');
  });

  test('Correctly adds and removes filter', () => {
    expect(filterSet.filters.length).toBe(1);
    const userInputFilter = new Filter({userInputContains: true, userInputValue: 'love'} as Filter);
    const checkboxFilter = new Filter({filterType: 'checkbox', filterValue: 'isPosed'} as Filter);

    filterSet.addFilter(userInputFilter);
    expect(filterSet.filters.length).toBe(2);

    filterSet.addFilter(checkboxFilter);
    expect(filterSet.filters.length).toBe(3);
    expect(filterSet).toMatchSnapshot();

    filterSet.removeFilter(userInputFilter)
    expect(filterSet.filters.length).toBe(2);
    expect(filterSet).toMatchSnapshot();
  });

  test('Correctly creates AND search term from multiple filters', () => {
    // Fill out default first filter
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('love');

    // Add additional posed filter
    const checkboxFilter = new Filter({filterType: 'checkbox', filterValue: 'isPosed'} as Filter);
    filterSet.addFilter(checkboxFilter);

    expect(filterSet.toString()).toBe('+name:*_love_* +isPosed:true')
  });
});

describe('Filter', () => {
  test('Initializes to with correct default data', () => {
    let filter = new Filter();
    expect(filter.filterType).toBe('userInput');
    expect(filter.filterValue).toBe('name');
    expect(filter.userInputValue).toBe('');
    expect(filter.userInputContains).toBe(true);
    expect(filter.selectType).toBe('only');
    expect(filter.selections.length).toBe(0);
    expect(filter.checkboxIsChecked).toBe(true);
    expect(filter).toMatchSnapshot();
  });

  test('Setters work correctly', () => {
    let filter = new Filter();

    filter.setFilterType('select');
    filter.setFilterValue('genre');
    filter.setUserInputValue('test-string');
    filter.setUserInputContains(false);
    filter.setSelectType('any');
    filter.setSelection(['a']);
    filter.setCheckboxIsChecked(false);

    expect(filter.filterType).toBe('select');
    expect(filter.filterValue).toBe('genre');
    expect(filter.userInputValue).toBe('test-string');
    expect(filter.userInputContains).toBe(false);
    expect(filter.selectType).toBe('any');
    expect(filter.selections.length).toBe(1);
    expect(filter.checkboxIsChecked).toBe(false);
    expect(filter).toMatchSnapshot();
  });

  test('User input filter returns correct search strings', () => {
    let userInputFilter = new Filter({
        userInputContains: true,
        userInputValue: 'New Love'
      } as Filter);
    expect(userInputFilter.toString()).toBe('+name:*_new_love_*')

    userInputFilter = new Filter({
      userInputContains: false,
      userInputValue: 'New Love'
    } as Filter);
    expect(userInputFilter.toString()).toBe('-name:*_new_love_*')

    userInputFilter = new Filter({
      userInputContains: true,
      userInputValue: 'Brilliant Light ♥ Rapunzel'
    } as Filter);
    expect(userInputFilter.toString()).toBe('+name:*_brilliant_light_♥_rapunzel_*')   
  });

  test('Select filter returns correct search strings', () => {
    let selectFilter = new Filter({
      filterType: 'select',
      filterValue: 'genre',
      selectType: 'only',
      selections: ['14']
    } as Filter);
    expect(selectFilter.toString()).toBe('+genre:14')

    selectFilter = new Filter({
      filterType: 'select',
      filterValue: 'special',
      selectType: 'any',
      selections: ['2', '4', '6', '8']
    } as Filter);
    expect(selectFilter.toString()).toBe(undefined)
  });

  test('Checkbox filter returns correct search strings', () => {
    let checkboxFilter = new Filter({
        filterType: 'checkbox',
        filterValue: 'isSuit',
        checkboxIsChecked: true,
      } as Filter);
    expect(checkboxFilter.toString()).toBe('+isSuit:true')

    checkboxFilter = new Filter({
      filterType: 'checkbox',
      filterValue: 'isPosed',
      checkboxIsChecked: false,
    } as Filter);
    expect(checkboxFilter.toString()).toBe('+isPosed:false')
  });
});

describe('FilterState', () => {
  let store: Store<any>;
  let state: RootState;
  let mockRootReducer: any;
  
  beforeEach(() => {
    mockRootReducer = combineReducers({
      search: searchReducer,
    });
    store = createStoreWithMiddleware(mockRootReducer);
    store.dispatch<any>(updateFilterSet(new FilterSet()));
  });

  test('Assert initial state uses default values', async () => {
    const filterSet: FilterSet = store.getState().search.filterSet;
    expect(filterSet.filters.length).toEqual(1);
    expect(filterSet.operator).toBe('and');
  });

  test(`If there are filters, ignore simple search`, async () => {
    // Fill out default first filter
    const filterSet: FilterSet = store.getState().search.filterSet;
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('ariel');

    await store.dispatch<any>(setAdvancedSearch(true));
    await store.dispatch<any>(updateSearchString('love'));
    await store.dispatch<any>(updateMaxResults(3));
    await store.dispatch<any>(searchInventory());
    state = store.getState();
    expect(state.search.results[0].displayName.includes('Ariel')).toBe(true);
  });

  test(`Looking up items via the NAME filter - ONLY items with "ariel" included`, async () => {
    // Fill out default first filter
    const filterSet: FilterSet = store.getState().search.filterSet;
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('ariel');

    await store.dispatch<any>(setAdvancedSearch(true));
    await store.dispatch<any>(updateMaxResults(20));
    await store.dispatch<any>(searchInventory());

    state = store.getState();
    expect(state.search.results.length).toEqual(16);
    expect(state.search.filterSet).toMatchSnapshot();
    expect(state.search.results).toMatchSnapshot();
  });

  test(`Looking up items via the NAME filter - ONLY items with "Legend of Light" included`, async () => {
    // Fill out default first filter
    const filterSet: FilterSet = store.getState().search.filterSet;
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('Legend of Light');

    await store.dispatch<any>(setAdvancedSearch(true));
    await store.dispatch<any>(updateMaxResults(3));
    await store.dispatch<any>(searchInventory());
    state = store.getState();
    expect(state.search.results[0].displayName.includes('Legend of Light')).toBe(true);

    expect(state.search.filterSet).toMatchSnapshot();
    expect(state.search.results).toMatchSnapshot();
  });

  test(`OR filter - 'little mermaid' OR 'legend of light`, async () => {
    // Fill out default first filter
    const filterSet: FilterSet = store.getState().search.filterSet;
    filterSet.setOperator('or');
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('Legend of Light');

    // Add second filter
    const userInputFilter2 = new Filter({
      userInputContains: true,
      userInputValue: 'Little Mermaid'
    } as Filter);
    filterSet.addFilter(userInputFilter2);

    await store.dispatch<any>(setAdvancedSearch(true));
    await store.dispatch<any>(updateMaxResults(10));
    await store.dispatch<any>(searchInventory());
    state = store.getState();
    
    // 2 mermaid suits, 2 LoL suits, 1 LoL dress
    expect(state.search.results.length).toEqual(5); 
    expect(state.search.filterSet).toMatchSnapshot();
    expect(state.search.results).toMatchSnapshot();
  });

  test(`AND filter - 'little mermaid' AND 'legend of light`, async () => {
    // Fill out default first filter
    const filterSet: FilterSet = store.getState().search.filterSet;
    filterSet.setOperator('and');
    filterSet.filters[0].setUserInputContains(true);
    filterSet.filters[0].setUserInputValue('Legend of Light');

    // Add second filter
    const userInputFilter2 = new Filter({
      userInputContains: true,
      userInputValue: 'Little Mermaid'
    } as Filter);
    filterSet.addFilter(userInputFilter2);

    await store.dispatch<any>(setAdvancedSearch(true));
    await store.dispatch<any>(updateMaxResults(10));
    await store.dispatch<any>(searchInventory());
    state = store.getState();

    expect(state.search.results.length).toEqual(0); 
    expect(state.search.filterSet).toMatchSnapshot();
    expect(state.search.results).toMatchSnapshot();
  });
});
