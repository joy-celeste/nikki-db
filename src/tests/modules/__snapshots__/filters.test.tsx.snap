// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Initializes to with correct default data 1`] = `
Filter {
  "checkboxIsChecked": true,
  "filterType": "userInput",
  "filterValue": "name",
  "id": "filter-aoeugk",
  "selectType": "only",
  "selections": Array [],
  "userInputContains": true,
  "userInputValue": "",
}
`;

exports[`Filter Setters work correctly 1`] = `
Filter {
  "checkboxIsChecked": false,
  "filterType": "select",
  "filterValue": "genre",
  "id": "filter-6m3a49",
  "selectType": "any",
  "selections": Array [
    "a",
  ],
  "userInputContains": false,
  "userInputValue": "test-string",
}
`;

exports[`FilterSet Correctly adds and removes filter 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-qilp03",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "",
    },
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-qqm5x3",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "love",
    },
    Filter {
      "checkboxIsChecked": true,
      "filterType": "checkbox",
      "filterValue": "isPosed",
      "id": "filter-4e4b9",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "",
    },
  ],
  "id": "filterSet-i9qlh",
  "operator": "and",
}
`;

exports[`FilterSet Correctly adds and removes filter 2`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-qilp03",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "",
    },
    Filter {
      "checkboxIsChecked": true,
      "filterType": "checkbox",
      "filterValue": "isPosed",
      "id": "filter-4e4b9",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "",
    },
  ],
  "id": "filterSet-i9qlh",
  "operator": "and",
}
`;

exports[`FilterSet Initializes to with correct default data 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-7qhq1",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "",
    },
  ],
  "id": "filterSet-ny18ce",
  "operator": "and",
}
`;

exports[`FilterState AND filter - 'little mermaid' AND 'legend of light 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-2h8ni",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "Legend of Light",
    },
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-lc16bw",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "Little Mermaid",
    },
  ],
  "id": "filterSet-wbvgrl",
  "operator": "and",
}
`;

exports[`FilterState AND filter - 'little mermaid' AND 'legend of light 2`] = `Array []`;

exports[`FilterState Looking up items via the NAME filter - ONLY items with "Legend of Light" included 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-ad3wrr",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "Legend of Light",
    },
  ],
  "id": "filterSet-3q0d2c",
  "operator": "and",
}
`;

exports[`FilterState Looking up items via the NAME filter - ONLY items with "Legend of Light" included 2`] = `
Array [
  Object {
    "contents": Array [
      60984,
      71980,
      22025,
      30996,
      86912,
      86913,
      86910,
      86911,
      90429,
      12277,
      86915,
      86914,
    ],
    "displayName": "The Legend of Light (Suit)",
    "iconId": 22025,
    "isSuit": true,
    "key": "22025-The Legend of Light (Suit)-0",
    "posed": false,
    "variation": "0",
  },
  Object {
    "contents": Array [
      86911,
      86912,
      22026,
      86910,
      86918,
      90429,
      86916,
      86917,
      12278,
    ],
    "displayName": "The Legend of Light (Posed Suit)",
    "iconId": 22026,
    "isSuit": true,
    "key": "22026-The Legend of Light (Posed Suit)-1",
    "posed": true,
    "variation": "1",
  },
  Object {
    "contents": Array [
      "22026",
    ],
    "displayName": "The Legend of Light",
    "iconId": "22026",
    "key": "22026-The Legend of Light",
  },
]
`;

exports[`FilterState Looking up items via the NAME filter - ONLY items with "ariel" included 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-9zkqva",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "ariel",
    },
  ],
  "id": "filterSet-hcyj76",
  "operator": "and",
}
`;

exports[`FilterState Looking up items via the NAME filter - ONLY items with "ariel" included 2`] = `
Array [
  Object {
    "contents": Array [
      "90342",
    ],
    "displayName": "Yearning ♥ Ariel",
    "iconId": "90342",
    "key": "90342-Yearning ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85622",
    ],
    "displayName": "Treasure ♥ Ariel",
    "iconId": "85622",
    "key": "85622-Treasure ♥ Ariel",
  },
  Object {
    "contents": Array [
      "21771",
    ],
    "displayName": "Travel ♥ Ariel",
    "iconId": "21771",
    "key": "21771-Travel ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85627",
    ],
    "displayName": "Strong Soul ♥ Ariel",
    "iconId": "85627",
    "key": "85627-Strong Soul ♥ Ariel",
  },
  Object {
    "contents": Array [
      "40766",
    ],
    "displayName": "Slumber ♥ Ariel",
    "iconId": "40766",
    "key": "40766-Slumber ♥ Ariel",
  },
  Object {
    "contents": Array [
      "11990",
    ],
    "displayName": "Purity ♥ Ariel",
    "iconId": "11990",
    "key": "11990-Purity ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85619",
    ],
    "displayName": "Ocean Crown ♥ Ariel",
    "iconId": "85619",
    "key": "85619-Ocean Crown ♥ Ariel",
  },
  Object {
    "contents": Array [
      "71747",
    ],
    "displayName": "Light ♥ Ariel",
    "iconId": "71747",
    "key": "71747-Light ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85628",
    ],
    "displayName": "Innocent Heart ♥ Ariel",
    "iconId": "85628",
    "key": "85628-Innocent Heart ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85626",
    ],
    "displayName": "Ice Soul ♥ Ariel",
    "iconId": "85626",
    "key": "85626-Ice Soul ♥ Ariel",
  },
  Object {
    "contents": Array [
      "50707",
    ],
    "displayName": "Hope ♥ Ariel",
    "iconId": "50707",
    "key": "50707-Hope ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85624",
    ],
    "displayName": "Free ♥ Ariel",
    "iconId": "85624",
    "key": "85624-Free ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85623",
    ],
    "displayName": "Flower Stone ♥ Ariel",
    "iconId": "85623",
    "key": "85623-Flower Stone ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85625",
    ],
    "displayName": "Colorful ♥ Ariel",
    "iconId": "85625",
    "key": "85625-Colorful ♥ Ariel",
  },
  Object {
    "contents": Array [
      "85621",
    ],
    "displayName": "Carefree ♥ Ariel",
    "iconId": "85621",
    "key": "85621-Carefree ♥ Ariel",
  },
  Object {
    "contents": Array [
      "11989",
    ],
    "displayName": "Ardor ♥ Ariel",
    "iconId": "11989",
    "key": "11989-Ardor ♥ Ariel",
  },
]
`;

exports[`FilterState OR filter - 'little mermaid' OR 'legend of light 1`] = `
FilterSet {
  "filters": Array [
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-vo29lj",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "Legend of Light",
    },
    Filter {
      "checkboxIsChecked": true,
      "filterType": "userInput",
      "filterValue": "name",
      "id": "filter-n7upac",
      "selectType": "only",
      "selections": Array [],
      "userInputContains": true,
      "userInputValue": "Little Mermaid",
    },
  ],
  "id": "filterSet-jt4o0b",
  "operator": "or",
}
`;

exports[`FilterState OR filter - 'little mermaid' OR 'legend of light 2`] = `
Array [
  Object {
    "contents": Array [
      60984,
      71980,
      22025,
      30996,
      86912,
      86913,
      86910,
      86911,
      90429,
      12277,
      86915,
      86914,
    ],
    "displayName": "The Legend of Light (Suit)",
    "iconId": 22025,
    "isSuit": true,
    "key": "22025-The Legend of Light (Suit)-0",
    "posed": false,
    "variation": "0",
  },
  Object {
    "contents": Array [
      86911,
      86912,
      22026,
      86910,
      86918,
      90429,
      86916,
      86917,
      12278,
    ],
    "displayName": "The Legend of Light (Posed Suit)",
    "iconId": 22026,
    "isSuit": true,
    "key": "22026-The Legend of Light (Posed Suit)-1",
    "posed": true,
    "variation": "1",
  },
  Object {
    "contents": Array [
      "22026",
    ],
    "displayName": "The Legend of Light",
    "iconId": "22026",
    "key": "22026-The Legend of Light",
  },
  Object {
    "contents": Array [
      71747,
      85619,
      40766,
      50707,
      85622,
      85623,
      85620,
      85621,
      85626,
      85627,
      90342,
      11989,
      85625,
      85624,
    ],
    "displayName": "The Little Mermaid (Suit)",
    "iconId": 40766,
    "isSuit": true,
    "key": "40766-The Little Mermaid (Suit)-0",
    "posed": false,
    "variation": "0",
  },
  Object {
    "contents": Array [
      85623,
      85625,
      21771,
      85619,
      90342,
      11990,
      85626,
      85627,
    ],
    "displayName": "The Little Mermaid (Posed Suit)",
    "iconId": 21771,
    "isSuit": true,
    "key": "21771-The Little Mermaid (Posed Suit)-1",
    "posed": true,
    "variation": "1",
  },
]
`;
